identity-android/src/main/java/com/android/identity/android/direct_access/DirectAccess.kt

```diff
@@ -0,0 +1,493 @@
+        val bos = ByteArrayOutputStream()
+        bos.write(0) // CLS
+        bos.write(INS_ENVELOPE.toInt()) // INS
+        bos.write(0) // P1
+        bos.write(0) // P2
+        // Send extended length APDU always as response size is not known to HAL.
+        // Case 1: Lc > 0  CLS | INS | P1 | P2 | 00 | 2 bytes of Lc | CommandData | 2 bytes of Le
+        // all set to 00.
+        // Case 2: Lc = 0  CLS | INS | P1 | P2 | 3 bytes of Le all set to 00.
+        bos.write(0x00)
+        // Extended length 3 bytes, starts with 0x00
+        if (data.isNotEmpty()) {
+            bos.write(data.size shr 8)
+            bos.write(data.size and 0xFF)
+            // Data
+            bos.write(data)
+        }
+        bos.write(0)
+        bos.write(0)
+        return bos.toByteArray()
+    }
+
+    @Throws(IOException::class)
+    private fun sendApdu(
+        cmd: Int,
+        slot: Int,
+        data: ByteArray,
+        offset: Int,
+        length: Int,
+        operation: Byte,
+    ): ByteArray? {
+        val bb = ByteBuffer.allocate(length)
+        bb.put(data, offset, length)
+        val scratchpad = ByteArray(2)
+        val bos = ByteArrayOutputStream()
+        // set instruction
+        setShort(scratchpad, 0, cmd.toShort())
+        bos.write(scratchpad)
+        bos.write(slot)
+        bos.write(operation.toInt())
+        bos.write(Cbor.encode((Bstr(bb.array()))))
+        val beginApdu = makeCommandApdu(bos.toByteArray())
+
+        val response: ByteArray = transport.sendData(beginApdu)
+        val status = getAPDUResponseStatus(response)
+        check(APDU_RESPONSE_STATUS_OK == status) {
+            "Operation failed. Response status: $status" }
+        if (response.size > 2) {
+            val input = response.copyOf(response.size - 2)
+            return decode(input).asBstr
+        }
+        return null
+    }
+
+    /**
+     * Initializes the transport so that the app can begin connecting to the applet in the background.
+     */
+    fun warmupTransport() {
+        DirectAccessOmapiTransport()
+    }
+
+    /**
+     * Allocates space (a slot) in the Direct Access applet for a
+     * document to be provisioned to.
+     *
+     * @return ID of the allocated slot, -1 if no slot is available
+     */
+    fun allocateDocumentSlot(): Int {
+        val apdu: ByteArray?
+        val response: ByteArray?
+        try {
+            transport.closeConnection()
+            transport.openConnection()
+
+            val scratchpad = ByteArray(2)
+            val bos = ByteArrayOutputStream()
+            // set instruction
+            setShort(scratchpad, 0, CMD_MDOC_CREATE.toShort())
+            bos.write(scratchpad)
+            apdu = makeCommandApdu(bos.toByteArray())
+
+            response = transport.sendData(apdu)
+            check(getAPDUResponseStatus(response) == APDU_RESPONSE_STATUS_OK)
+            transport.closeConnection()
+            return response[0].toInt()
```

Comment:
Seems like we're also expecting the response length to be 3 bytes? Should we check for that? Or at least log if it's an unexpected length?

Reply:
Agreed. We will add this check.